File: app/src/main/java/com/stocky/Adapters/ProductsAdapter.java
package com.stocky.Adapters;

import android.app.AlertDialog;
import android.content.DialogInterface;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.TextView;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import com.google.android.gms.tasks.OnSuccessListener;
import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseError;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.FirebaseDatabase;
import com.google.firebase.database.Query;
import com.google.firebase.database.ValueEventListener;
import com.stocky.Model.Product;
import com.stocky.R;

import java.util.ArrayList;
import java.util.HashMap;

public class ProductsAdapter extends RecyclerView.Adapter<ProductsAdapter.ViewHolder> {

    private final ArrayList<Product> products;

    public ProductsAdapter(ArrayList<Product> products) {
        this.products = products;
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View v = inflater.inflate(R.layout.item_produto, parent, false);
        ViewHolder vh = new ViewHolder(v);
        return vh;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        final String product = products.get(position).getDescricao();
        holder.nomeProduto.setText(product);
        holder.nomeProduto.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Product productToUpdateName = products.get(position);
                EditText textNewProductName = new EditText(v.getContext());
                textNewProductName.setHint("Digite o nome do produto");
                textNewProductName.setPadding(16, 16, 16, 16);
                new AlertDialog.Builder(v.getContext())
                        .setTitle("ALTERAR PRODUTO")
                        .setMessage("Alterar nome do produto " + productToUpdateName.getDescricao())
                        .setCancelable(false)
                        .setView(textNewProductName)
                        .setPositiveButton("SALVAR", new DialogInterface.OnClickListener() {
                            @Override
                            public void onClick(DialogInterface dialog, int which) {
                                String id = productToUpdateName.getId();
                                String descricao = productToUpdateName.getDescricao();
                                String newProductText = textNewProductName.getText().toString();
                                HashMap hashMapProductDesc = new HashMap();
                                HashMap hashMapHistoryProd = new HashMap();
                                hashMapProductDesc.put("descricao", newProductText);
                                DatabaseReference databaseProducts = FirebaseDatabase.getInstance().getReference("products").child(id);
                                DatabaseReference databaseHistory = FirebaseDatabase.getInstance().getReference("history");
                                Query historyQuery = databaseHistory.orderByChild("product").equalTo(descricao);
                                databaseProducts.updateChildren(hashMapProductDesc).addOnSuccessListener(new OnSuccessListener() {
                                    @Override
                                    public void onSuccess(Object o) {
                                        historyQuery.addListenerForSingleValueEvent(new ValueEventListener() {
                                            @Override
                                            public void onDataChange(@NonNull DataSnapshot snapshot) {
                                                for (DataSnapshot historySnapshot : snapshot.getChildren()) {
                                                    hashMapHistoryProd.put(historySnapshot.getKey()+"/product", newProductText);

                                                }
                                                databaseHistory.updateChildren(hashMapHistoryProd);
                                            }

                                            @Override
                                            public void onCancelled(@NonNull DatabaseError error) {

                                            }
                                        });
                                    }
                                });
                            }
                        })
                        .setNegativeButton("CANCELAR", null).show();
            }
        });
        holder.btnRemoveProd.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Product productToDelete = products.get(position);

                new AlertDialog.Builder(v.getContext())
                        .setTitle("REMOVER PRODUTO")
                        .setMessage("Tem certeza que deseja remover o produto " + productToDelete.getDescricao() + "?")
                        .setCancelable(false)
                        .setPositiveButton("SIM", new DialogInterface.OnClickListener() {

                            @Override
                            public void onClick(DialogInterface dialogInterface, int i) {

                                String id = productToDelete.getId();
                                String descricao = productToDelete.getDescricao();

                                DatabaseReference databaseProducts = FirebaseDatabase.getInstance().getReference("products").child(id);
                                DatabaseReference databaseHistory = FirebaseDatabase.getInstance().getReference("history");

                                Query historyQuery = databaseHistory.orderByChild("product").equalTo(descricao);

                                databaseProducts.removeValue();

                                historyQuery.addListenerForSingleValueEvent(new ValueEventListener() {
                                    @Override
                                    public void onDataChange(@NonNull DataSnapshot snapshot) {
                                        for (DataSnapshot historySnapshot : snapshot.getChildren()) {
                                            historySnapshot.getRef().removeValue();
                                        }
                                    }

                                    @Override
                                    public void onCancelled(@NonNull DatabaseError error) {

                                    }
                                });

                                Toast.makeText(holder.btnRemoveProd.getContext(), "O produto foi removido.", Toast.LENGTH_SHORT).show();
                            } }).setNegativeButton("NÃO", null).show();
            }
        });
    }

    @Override
    public int getItemCount() {
        return products.size();
    }

    public static class ViewHolder extends RecyclerView.ViewHolder {
        public TextView nomeProduto;
        public View layout;
        public Button btnRemoveProd;

        public ViewHolder(View itemView) {
            super(itemView);
            layout = itemView;
            nomeProduto = itemView.findViewById(R.id.nomeProduto);
            btnRemoveProd = itemView.findViewById(R.id.btnRemoveProd);
        }
    }
}

File: app/src/main/java/com/stocky/adapters/ProductsAdapter.kt
package com.stocky.adapters

import com.stocky.model.Product
import androidx.recyclerview.widget.RecyclerView
import android.view.ViewGroup
import android.view.LayoutInflater
import android.view.View
import com.stocky.R
import android.widget.EditText
import com.google.firebase.database.FirebaseDatabase
import com.google.firebase.database.ValueEventListener
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseError
import android.widget.Toast
import android.widget.TextView
import androidx.constraintlayout.widget.ConstraintLayout
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import java.util.ArrayList
import java.util.HashMap

class ProductsAdapter(private val products: ArrayList<Product?>) :
    RecyclerView.Adapter<ProductsAdapter.ViewHolder>() {
    override fun onCreateViewHolder(
        parent: ViewGroup,
        viewType: Int
    ): ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val v = inflater.inflate(R.layout.item_product, parent, false)
        return ViewHolder(v)
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        val product = products[position]?.descricao
        val qtd = products[position]?.qtdEstoque
        val dateChange = products[position]?.dt_alteracao
        holder.productName.text = product
        holder.stockQuantity.text = holder.stockQuantity.context.getString(R.string.product_quantity, qtd.toString())
        holder.dateChange.text = holder.dateChange.context.getString(R.string.product_date_change, dateChange)
        holder.productView.setOnClickListener { v ->
            val productToUpdateName = products[position]
            val textNewProductName = EditText(v.context)
            textNewProductName.hint = v.context.getString(R.string.type_new_product_name)
            textNewProductName.setPadding(16, 16, 16, 16)
            MaterialAlertDialogBuilder(v.context)
                .setTitle(v.context.getString(R.string.change_product))
                .setIcon(R.drawable.ic_warning)
                .setMessage(v.context.getString(R.string.change_product_description, productToUpdateName?.descricao))
                .setCancelable(false)
                .setView(textNewProductName)

                .setPositiveButton(v.context.getString(R.string.save)) { _, _ ->
                    val id = productToUpdateName?.id
                    val description = productToUpdateName?.descricao
                    val newProductText = textNewProductName.text.toString()
                    val hashMapProductDesc: HashMap<String, Any> = HashMap<String, Any>()
                    val hashMapHistoryProd: HashMap<String, Any> = HashMap<String, Any>()
                    hashMapProductDesc["descricao"] = newProductText
                    val databaseProducts =
                        FirebaseDatabase.getInstance().getReference("products").child(
                            id!!
                        )
                    val databaseHistory = FirebaseDatabase.getInstance().getReference("history")
                    val historyQuery = databaseHistory.orderByChild("product").equalTo(description)
                    databaseProducts.updateChildren(hashMapProductDesc).addOnSuccessListener {
                        historyQuery.addListenerForSingleValueEvent(object : ValueEventListener {
                            override fun onDataChange(snapshot: DataSnapshot) {
                                for (historySnapshot in snapshot.children) {
                                    hashMapHistoryProd[historySnapshot.key + "/product"] =
                                        newProductText
                                }
                                databaseHistory.updateChildren(hashMapHistoryProd)
                            }

                            override fun onCancelled(error: DatabaseError) {}
                        })
                    }
                }

                .setNegativeButton(v.context.getString(R.string.cancel), null).show()
        }
        holder.btnRemoveProduct.setOnClickListener { v ->
            val productToDelete = products[position]
            MaterialAlertDialogBuilder(v.context)
                .setTitle(v.context.getString(R.string.delete_product))
                .setIcon(R.drawable.ic_warning)
                .setMessage(v.context.getString(R.string.confirm_delete_product, productToDelete?.descricao))
                .setCancelable(false)
                .setPositiveButton(v.context.getString(R.string.yes)) { _, _ ->
                    val id = productToDelete?.id
                    val description = productToDelete?.descricao
                    val databaseProducts =
                        FirebaseDatabase.getInstance().getReference("products").child(
                            id!!
                        )
                    val databaseHistory = FirebaseDatabase.getInstance().getReference("history")
                    val historyQuery = databaseHistory.orderByChild("product").equalTo(description)
                    databaseProducts.removeValue()
                    historyQuery.addListenerForSingleValueEvent(object : ValueEventListener {
                        override fun onDataChange(snapshot: DataSnapshot) {
                            for (historySnapshot in snapshot.children) {
                                historySnapshot.ref.removeValue()
                            }
                        }

                        override fun onCancelled(error: DatabaseError) {}
                    })
                    Toast.makeText(
                        v.context,
                        v.context.getString(R.string.product_deleted),
                        Toast.LENGTH_SHORT
                    ).show()
                }.setNegativeButton(v.context.getString(R.string.no), null).show()
        }
    }

    override fun getItemCount(): Int {
        return products.size
    }

    class ViewHolder(var layout: View) : RecyclerView.ViewHolder(
        layout
    ) {
        var productName: TextView = itemView.findViewById(R.id.productName)
        var btnRemoveProduct: com.google.android.material.imageview.ShapeableImageView = itemView.findViewById(R.id.btnRemoveProduct)
        var stockQuantity : TextView = itemView.findViewById(R.id.stockQuantity)
        var dateChange : TextView = itemView.findViewById(R.id.dateChange)
        var productView: ConstraintLayout = itemView.findViewById(R.id.product_view)
    }
}